processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -1371,6 +1361,77 @@ public Grouper.BufferComparator bufferComparatorWithAggregators(
       );
     }
 
+    @Override
+    public ObjectMapper decorateObjectMapper(ObjectMapper spillMapper)
+    {
+
+      final JsonDeserializer<RowBasedKey> deserializer = new JsonDeserializer<RowBasedKey>()
+      {
+        @Override
+        public RowBasedKey deserialize(
+            JsonParser jp,
+            DeserializationContext deserializationContext
+        ) throws IOException
+        {
+          if (!jp.isExpectedStartArrayToken()) {
+            throw DruidException.defensive("Expected array start token, received [%s]", jp.getCurrentToken());
+          }
+          jp.nextToken();
+
+          final ObjectCodec codec = jp.getCodec();
+          final int timestampAdjustment = includeTimestamp ? 1 : 0;
+          final int dimsToRead = timestampAdjustment + serdeHelpers.length;
+          int dimsReadSoFar = 0;
+          final Object[] objects = new Object[dimsToRead];
+
+          while (jp.currentToken() != JsonToken.END_ARRAY) {
+            if (dimsReadSoFar >= dimsToRead) {
+              throw DruidException.defensive("More dimensions encountered than expected [%d]", dimsToRead);
+            }
+
+            if (includeTimestamp && dimsReadSoFar == 0) {
+              // Read the timestamp
+              objects[dimsReadSoFar] = codec.readValue(jp, Long.class);
+            } else {
```

Comment:
it might be nice to pull this out of the while loop?

```

          if (includeTimestamp && jp.currentToken() != JsonToken.END_ARRAY) {
            objects[dimsReadSoFar++] = codec.readValue(jp, Long.class);
            jp.nextToken();
          }

          while (jp.currentToken() != JsonToken.END_ARRAY) {
            if (dimsReadSoFar >= dimsToRead) {
              throw DruidException.defensive("More dimensions encountered than expected [%d]", dimsToRead);
            }

            DruidException.conditionalDefensive(
                dimsReadSoFar < dimsToRead,
                "Insufficient serde helpers present"
            );
            // Read the dimension
            if (serdeHelpers[dimsReadSoFar - timestampAdjustment].getComplexClazz() == null) {
              objects[dimsReadSoFar] = codec.readValue(jp, Object.class);
              if (objects[dimsReadSoFar] instanceof Integer) {
                objects[dimsReadSoFar] = ((Integer) objects[dimsReadSoFar]).longValue();
              } else if (objects[dimsReadSoFar] instanceof Double) {
                objects[dimsReadSoFar] = ((Double) objects[dimsReadSoFar]).floatValue();
              }
            } else {
              objects[dimsReadSoFar] =
                  codec.readValue(jp, serdeHelpers[dimsReadSoFar - timestampAdjustment].getComplexClazz());
            }
            ++dimsReadSoFar;
            jp.nextToken();
          }
```

Reply:
I tried this, but the complexity is the same, given that we'd have to duplicate the defensive checks, loop counter increment and json stream increment. I have left it as is for now.

Reply:
i guess my thought was that we would be removing one conditional check of the loop of every column of every row which could add up, but i haven't really measured one way or another. This is a pretty hot loop though so worth thinking about making it as efficient as possible

Reply:
I was looking at it from code's complexity perspective. Your approach is better to make it more efficient. I'd guess most of the time branch predictions would make it quite equivalent. I'll make the changes to make it explicit however.